File: projects/basis/sdk-review/py-v2/basis/modules/agent_identity.py
import json
import base64
import logging
from typing import Optional, Dict, Any
from web3 import Web3
from .factory import load_abi

logger = logging.getLogger(__name__)

# ERC-8004 Identity Registry on BSC
IDENTITY_REGISTRY = "0x8004A169FB4a3325136EB29fA0ceB6D2e539a432"

# Inline ABI for the Identity Registry
IDENTITY_ABI = [
    {"inputs":[{"name":"agentURI","type":"string"}],"name":"register","outputs":[{"name":"agentId","type":"uint256"}],"stateMutability":"nonpayable","type":"function"},
    {"inputs":[],"name":"register","outputs":[{"name":"agentId","type":"uint256"}],"stateMutability":"nonpayable","type":"function"},
    {"inputs":[{"name":"owner","type":"address"}],"name":"balanceOf","outputs":[{"name":"","type":"uint256"}],"stateMutability":"view","type":"function"},
    {"inputs":[{"name":"tokenId","type":"uint256"}],"name":"ownerOf","outputs":[{"name":"","type":"address"}],"stateMutability":"view","type":"function"},
    {"inputs":[{"name":"agentId","type":"uint256"}],"name":"getAgentWallet","outputs":[{"name":"","type":"address"}],"stateMutability":"view","type":"function"},
    {"inputs":[{"name":"agentId","type":"uint256"},{"name":"metadataKey","type":"string"}],"name":"getMetadata","outputs":[{"name":"","type":"bytes"}],"stateMutability":"view","type":"function"},
    {"inputs":[{"name":"agentId","type":"uint256"},{"name":"metadataKey","type":"string"},{"name":"metadataValue","type":"bytes"}],"name":"setMetadata","outputs":[],"stateMutability":"nonpayable","type":"function"},
    {"inputs":[{"name":"agentId","type":"uint256"},{"name":"newURI","type":"string"}],"name":"setAgentURI","outputs":[],"stateMutability":"nonpayable","type":"function"},
    {"inputs":[{"name":"spender","type":"address"},{"name":"agentId","type":"uint256"}],"name":"isAuthorizedOrOwner","outputs":[{"name":"","type":"bool"}],"stateMutability":"view","type":"function"},
    {"inputs":[{"name":"agentId","type":"uint256"}],"name":"tokenURI","outputs":[{"name":"","type":"string"}],"stateMutability":"view","type":"function"},
    {"anonymous":False,"inputs":[{"indexed":True,"name":"agentId","type":"uint256"},{"indexed":False,"name":"agentURI","type":"string"},{"indexed":True,"name":"owner","type":"address"}],"name":"Registered","type":"event"},
]


class AgentIdentityModule:
    def __init__(self, client):
        self.client = client
        self.registry_address = Web3.to_checksum_address(IDENTITY_REGISTRY)
        self.contract = self.client.web3.eth.contract(
            address=self.registry_address, abi=IDENTITY_ABI
        )

    def _sync_tx(self, tx_hash: str):
        """Sync tx to backend. Non-fatal on failure."""
        try:
            if not tx_hash.startswith("0x"):
                tx_hash = "0x" + tx_hash
            self.client.api.sync_transaction(tx_hash)
        except Exception as e:
            logger.warning("Sync warning: %s", e)

    def _build_metadata_uri(self, wallet: str, config: Optional[Dict[str, Any]] = None) -> str:
        """Build base64-encoded metadata URI for on-chain registration."""
        config = config or {}
        metadata = {
            "type": "https://eips.ethereum.org/EIPS/eip-8004#registration-v1",
            "name": config.get("name", "Basis Agent"),
            "description": config.get("description"),
            "image": config.get("image"),
            "website": "https://launchonbasis.com",
            "profile": f"https://launchonbasis.com/profile/{wallet}",
            "protocol": "basis",
            "active": True,
            "capabilities": config.get("capabilities", ["trading"]),
            "supportedTrust": ["reputation"],
        }
        json_str = json.dumps(metadata)
        b64 = base64.b64encode(json_str.encode()).decode()
        return f"data:application/json;base64,{b64}"

    def _build_and_send_tx(self, function_call):
        if not self.client.account:
            raise ValueError("Private key required.")
        tx = function_call.build_transaction({
            "from": self.client.account.address,
            "nonce": self.client.web3.eth.get_transaction_count(self.client.account.address),
        })
        signed_tx = self.client.web3.eth.account.sign_transaction(tx, private_key=self.client.account.key)
        tx_hash = self.client.web3.eth.send_raw_transaction(signed_tx.raw_transaction)
        receipt = self.client.web3.eth.wait_for_transaction_receipt(tx_hash)
        return {"hash": tx_hash.hex(), "receipt": receipt}

    def is_registered(self, wallet: str) -> bool:
        """Check if a wallet has an agent NFT on the Identity Registry."""
        checksum = Web3.to_checksum_address(wallet)
        balance = self.contract.functions.balanceOf(checksum).call()
        return balance > 0

    def register(self, config: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Register the current wallet as an ERC-8004 agent. Returns hash and agentId."""
        if not self.client.account:
            raise ValueError("Private key required to register as agent.")

        uri = self._build_metadata_uri(self.client.account.address, config)
        func = self.contract.functions.register(uri)
        result = self._build_and_send_tx(func)

        # Parse agentId from Registered event
        agent_id = 0
        receipt = result["receipt"]
        for log_entry in receipt.get("logs", []):
            addr = log_entry.get("address", "")
            if addr.lower() == self.registry_address.lower():
                topics = log_entry.get("topics", [])
                if len(topics) >= 2:
                    # topics[0] = event sig, topics[1] = indexed agentId
                    raw = topics[1]
                    if isinstance(raw, bytes):
                        agent_id = int.from_bytes(raw, "big")
                    else:
                        agent_id = int(raw, 16) if isinstance(raw, str) else int(raw)
                    break

        self._sync_tx(result['hash'])
        result["agentId"] = agent_id
        return result

    def register_and_sync(self, config: Optional[Dict[str, Any]] = None) -> int:
        """Full registration flow: check on-chain, register if needed, sync to API.
        Returns the agentId.
        """
        if not self.client.account:
            raise ValueError("Private key required.")

        address = self.client.account.address

        # Check if already registered
        if self.is_registered(address):
            try:
                api_agent = self.lookup_from_api(address)
                if api_agent and api_agent.get("isAgent"):
                    return api_agent["agent"]["agentId"]
            except Exception:
                pass
            return 0

        # Register on-chain
        result = self.register(config)
        agent_id = result["agentId"]

        # Sync to API
        try:
            self._sync_to_api(address, agent_id, config)
        except Exception as e:
            logger.warning("Agent API sync warning: %s", e)

        return agent_id

    def _sync_to_api(self, wallet: str, agent_id: int, config: Optional[Dict[str, Any]] = None):
        """Save registration to backend API."""
        import requests
        config = config or {}
        body = {
            "wallet": wallet,
            "agentId": agent_id,
            "name": config.get("name", "Basis Agent"),
            "description": config.get("description"),
        }
        # Use the API session (has SIWE cookie)
        self.client.api._session_request("POST", "/agents", json=body)

    def lookup_from_api(self, wallet: str) -> Optional[Dict[str, Any]]:
        """Look up an agent by wallet via the API."""
        try:
            import requests
            res = requests.get(f"{self.client.api_domain}/api/agents/{wallet}")
            if res.ok:
                return res.json()
        except Exception:
            pass
        return None

    def list_agents(self, page: int = 1, limit: int = 20) -> Dict[str, Any]:
        """List all registered agents via the API."""
        import requests
        res = requests.get(f"{self.client.api_domain}/api/agents?page={page}&limit={limit}")
        res.raise_for_status()
        return res.json()

    def get_agent_uri(self, agent_id: int) -> str:
        """Get the tokenURI for a registered agent (on-chain)."""
        return self.contract.functions.tokenURI(agent_id).call()

    def get_agent_wallet(self, agent_id: int) -> str:
        """Get the wallet linked to an agent ID (on-chain)."""
        return self.contract.functions.getAgentWallet(agent_id).call()

    def get_metadata(self, agent_id: int, key: str) -> bytes:
        """Get metadata for an agent by key (on-chain)."""
        return self.contract.functions.getMetadata(agent_id, key).call()

    def set_agent_uri(self, agent_id: int, new_uri: str) -> Dict[str, Any]:
        """Update the agent's URI (on-chain). Must be the owner."""
        func = self.contract.functions.setAgentURI(agent_id, new_uri)
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

File: projects/basis/sdk-review/js-v2/src/modules/AgentIdentity.ts
import { BasisClient } from '../BasisClient';
import IERC20Artifact from '../abis/IERC20.json';
import { Address, decodeEventLog } from 'viem';

// ERC-8004 Identity Registry on BSC
const IDENTITY_REGISTRY = '0x8004A169FB4a3325136EB29fA0ceB6D2e539a432' as Address;

// Inline ABI for the Identity Registry (external contract, no compiled artifact)
const identityRegistryAbi = [
  {"inputs":[{"name":"agentURI","type":"string"}],"name":"register","outputs":[{"name":"agentId","type":"uint256"}],"stateMutability":"nonpayable","type":"function"},
  {"inputs":[],"name":"register","outputs":[{"name":"agentId","type":"uint256"}],"stateMutability":"nonpayable","type":"function"},
  {"inputs":[{"name":"owner","type":"address"}],"name":"balanceOf","outputs":[{"name":"","type":"uint256"}],"stateMutability":"view","type":"function"},
  {"inputs":[{"name":"tokenId","type":"uint256"}],"name":"ownerOf","outputs":[{"name":"","type":"address"}],"stateMutability":"view","type":"function"},
  {"inputs":[{"name":"agentId","type":"uint256"}],"name":"getAgentWallet","outputs":[{"name":"","type":"address"}],"stateMutability":"view","type":"function"},
  {"inputs":[{"name":"agentId","type":"uint256"},{"name":"metadataKey","type":"string"}],"name":"getMetadata","outputs":[{"name":"","type":"bytes"}],"stateMutability":"view","type":"function"},
  {"inputs":[{"name":"agentId","type":"uint256"},{"name":"metadataKey","type":"string"},{"name":"metadataValue","type":"bytes"}],"name":"setMetadata","outputs":[],"stateMutability":"nonpayable","type":"function"},
  {"inputs":[{"name":"agentId","type":"uint256"},{"name":"newURI","type":"string"}],"name":"setAgentURI","outputs":[],"stateMutability":"nonpayable","type":"function"},
  {"inputs":[{"name":"spender","type":"address"},{"name":"agentId","type":"uint256"}],"name":"isAuthorizedOrOwner","outputs":[{"name":"","type":"bool"}],"stateMutability":"view","type":"function"},
  {"inputs":[{"name":"agentId","type":"uint256"}],"name":"tokenURI","outputs":[{"name":"","type":"string"}],"stateMutability":"view","type":"function"},
  {"anonymous":false,"inputs":[{"indexed":true,"name":"agentId","type":"uint256"},{"indexed":false,"name":"agentURI","type":"string"},{"indexed":true,"name":"owner","type":"address"}],"name":"Registered","type":"event"},
] as const;

export interface AgentConfig {
  name?: string;
  description?: string;
  image?: string;
  capabilities?: string[];
}

export class AgentIdentityModule {
  private client: BasisClient;
  private registryAddress: Address;

  constructor(client: BasisClient) {
    this.client = client;
    this.registryAddress = IDENTITY_REGISTRY;
  }

  private async _syncTx(txHash: string) {
    try {
      await this.client.api.syncTransaction(txHash);
    } catch (e: any) {
      console.warn('Sync warning:', e.message || e);
    }
  }

  /**
   * Build the on-chain metadata JSON for an agent.
   */
  private buildMetadataUri(wallet: string, config?: AgentConfig): string {
    const metadata = {
      type: 'https://eips.ethereum.org/EIPS/eip-8004#registration-v1',
      name: config?.name || 'Basis Agent',
      description: config?.description || null,
      image: config?.image || null,
      website: 'https://launchonbasis.com',
      profile: `https://launchonbasis.com/profile/${wallet}`,
      protocol: 'basis',
      active: true,
      capabilities: config?.capabilities || ['trading'],
      supportedTrust: ['reputation'],
    };
    const json = JSON.stringify(metadata);
    const base64 = Buffer.from(json).toString('base64');
    return `data:application/json;base64,${base64}`;
  }

  /**
   * Check if a wallet has registered an agent on the Identity Registry.
   */
  async isRegistered(wallet: Address): Promise<boolean> {
    const balance = await this.client.publicClient.readContract({
      address: this.registryAddress,
      abi: identityRegistryAbi,
      functionName: 'balanceOf',
      args: [wallet],
    }) as bigint;
    return balance > 0n;
  }

  /**
   * Register the current wallet as an ERC-8004 agent.
   * Returns the agentId.
   *
   * If already registered on-chain, returns null (check via isRegistered first).
   */
  async register(config?: AgentConfig): Promise<{ hash: string; agentId: bigint }> {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error('Wallet required to register as agent.');
    }

    const account = this.client.walletClient.account;
    const uri = this.buildMetadataUri(account.address, config);

    const { request } = await this.client.publicClient.simulateContract({
      account,
      address: this.registryAddress,
      abi: identityRegistryAbi,
      functionName: 'register',
      args: [uri],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);

    // Parse agentId from Registered event
    let agentId = 0n;
    for (const log of receipt.logs) {
      if (log.address.toLowerCase() === this.registryAddress.toLowerCase()) {
        try {
          const decoded = decodeEventLog({
            abi: identityRegistryAbi,
            data: log.data,
            topics: log.topics,
          });
          if (decoded.eventName === 'Registered') {
            agentId = (decoded.args as any).agentId;
            break;
          }
        } catch {}
      }
    }

    return { hash, agentId };
  }

  /**
   * Full registration flow:
   * 1. Check if already registered on-chain
   * 2. If not, register on-chain
   * 3. Save to backend API
   *
   * Returns the agentId.
   */
  async registerAndSync(config?: AgentConfig): Promise<bigint> {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error('Wallet required to register as agent.');
    }

    const address = this.client.walletClient.account.address;

    // Check if already registered on-chain
    const alreadyRegistered = await this.isRegistered(address);

    let agentId: bigint;

    if (alreadyRegistered) {
      // Try to load from API
      try {
        const apiAgent = await this.lookupFromApi(address);
        if (apiAgent && apiAgent.isAgent) {
          return BigInt(apiAgent.agent.agentId);
        }
      } catch {}
      // Can't determine agentId without API — return 0
      return 0n;
    }

    // Register on-chain
    const result = await this.register(config);
    agentId = result.agentId;

    // Sync to backend API
    try {
      await this.syncToApi(address, agentId, config);
    } catch (err) {
      console.warn('Agent API sync warning:', err instanceof Error ? err.message : err);
    }

    return agentId;
  }

  /**
   * Sync agent registration to the backend API.
   */
  private async syncToApi(wallet: string, agentId: bigint, config?: AgentConfig): Promise<void> {
    const cookie = this.client.sessionCookie;
    if (!cookie) return;

    const body = JSON.stringify({
      wallet,
      agentId: Number(agentId),
      name: config?.name || 'Basis Agent',
      description: config?.description || null,
    });

    const res = await fetch(`${this.client.apiDomain}/api/agents`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Cookie: cookie,
      },
      body,
    });

    if (!res.ok) {
      const errBody = await res.text().catch(() => '');
      throw new Error(`Agent sync failed [${res.status}]: ${errBody}`);
    }
  }

  /**
   * Look up an agent by wallet address via the API.
   */
  async lookupFromApi(wallet: string): Promise<{ isAgent: boolean; agent: any } | null> {
    try {
      const res = await fetch(`${this.client.apiDomain}/api/agents/${wallet}`);
      if (!res.ok) return null;
      return res.json();
    } catch {
      return null;
    }
  }

  /**
   * List all registered agents via the API.
   */
  async listAgents(page: number = 1, limit: number = 20): Promise<any> {
    const res = await fetch(`${this.client.apiDomain}/api/agents?page=${page}&limit=${limit}`);
    if (!res.ok) throw new Error(`Failed to list agents: ${res.status}`);
    return res.json();
  }

  /**
   * Get the tokenURI for a registered agent (on-chain).
   */
  async getAgentURI(agentId: bigint): Promise<string> {
    return this.client.publicClient.readContract({
      address: this.registryAddress,
      abi: identityRegistryAbi,
      functionName: 'tokenURI',
      args: [agentId],
    }) as Promise<string>;
  }

  /**
   * Get the wallet linked to an agent ID (on-chain).
   */
  async getAgentWallet(agentId: bigint): Promise<Address> {
    return this.client.publicClient.readContract({
      address: this.registryAddress,
      abi: identityRegistryAbi,
      functionName: 'getAgentWallet',
      args: [agentId],
    }) as Promise<Address>;
  }

  /**
   * Get metadata for an agent by key (on-chain).
   */
  async getMetadata(agentId: bigint, key: string): Promise<string> {
    return this.client.publicClient.readContract({
      address: this.registryAddress,
      abi: identityRegistryAbi,
      functionName: 'getMetadata',
      args: [agentId, key],
    }) as Promise<string>;
  }

  /**
   * Update the agent's URI (on-chain). Must be the owner.
   */
  async setAgentURI(agentId: bigint, newURI: string) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error('Wallet required.');
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.registryAddress,
      abi: identityRegistryAbi,
      functionName: 'setAgentURI',
      args: [agentId, newURI],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });
    this._syncTx(hash);
    return { hash, receipt };
  }
}
